### PR TITLE
Nemo file updates

### PR DESCRIPTION
**Pull request by okuchaiev**
This PR does the following:
* [x] add 'target' to conifgs for models
* [x] use torch to restore state dict instead of PTL
* [x] use torch to save state dict  instead of PTL
* [x] ability to ship artifacts inside .nemo file
* [x] cloud instantiation
* [x] asr inference script
* [x] CI WER accuracy test based on cloud instantiation an asr inference script

### CODE REVIEW

File: nemo/collections/nlp/models/token_classification/ner_model.py [hunk 1]
```diff
@@ -54,36 +54,59 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
             data_dir=cfg.data_dir, modes=["train", "test", "dev"], pad_label=cfg.pad_label
         )
         self.data_dir = cfg.data_dir
-        self.model_cfg = cfg
+        # self.model_cfg = cfg
```
Comment (okuchaiev) [inline nemo/collections/nlp/models/token_classification/ner_model.py:57]: @titu1994  and @ekmb what do you think about this?

Comment (titu1994) [reply nemo/collections/nlp/models/token_classification/ner_model.py:57]: I think this is good but slightly problematic? So the logic is:
 - If path exists, 
    - get its file basename
    - test again if file exists at ("./{basename}")? 

This is problematic cause we generally store vocabs somewhere else. So the second "if case" will fail unless we copy the vocab dir to the exact location where the code is run (which we can do, but more logic is needed).

Comment (titu1994) [reply nemo/collections/nlp/models/token_classification/ner_model.py:57]: Can we do a double check - the first check if the file is resolved, great, set `vocab_file={cfg.language_model.vocab_file}`.

If that check fails, then compute base name and check for second test (as in how the serialization code will store directly inside the current execution dir so this time it will succeed).

If both fail, raise file not found error?

Comment (titu1994) [reply nemo/collections/nlp/models/token_classification/ner_model.py:57]: After discussion, this looks like good logic.

Comment (okuchaiev) [reply nemo/collections/nlp/models/token_classification/ner_model.py:57]: refactored into separate function after discussion

Comment (ekmb) [reply nemo/collections/nlp/models/token_classification/ner_model.py:57]: @okuchaiev Sorry, missed the question. Why do you need tokenizer_model - it's only used for sentence piece. Most of the models work with Nemobert tokenizer and use only pretrained_model_name for init. I'd suggest initializing the tokenizer first and them pull the vocab out of it and store it in artifacts, maybe store vocab + tokenizer cls name, but not the model

Comment (okuchaiev) [reply nemo/collections/nlp/models/token_classification/ner_model.py:57]: @ekmb it stores whatever artifacts were provided using during model instantiation. So if vocab were provided it should store it, if tokenizer_model were provided it will store that. If either is None it will not add it to .nemo file on saving

File: Jenkinsfile [hunk 2]
```diff
@@ -160,7 +159,37 @@ pipeline {
       }
     }
     
+//     stage('L2 QuartzNet En Checkpoint inference Test' {
```
Comment (titu1994) [inline Jenkinsfile:162]: Is this going to be added later?

File: nemo/collections/nlp/models/token_classification/ner_model.py [hunk 3]
```diff
@@ -54,36 +54,40 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
 
         self.tokenizer = get_tokenizer(
             tokenizer_name=cfg.language_model.tokenizer,
             pretrained_model_name=cfg.language_model.pretrained_model_name,
-            vocab_file=cfg.language_model.vocab_file,
-            tokenizer_model=cfg.language_model.tokenizer_model,
+            vocab_file=self.register_artifact(
+                conf_path='language_model.vocab_file', src=cfg.language_model.vocab_file
+            ),
```
Comment (ekmb) [inline nemo/collections/nlp/models/token_classification/ner_model.py:64]: @okuchaiev In most cases, we don't provide a vocab file and use the default one pulled by means of the "pretrained_model_name". So, in this setup vocab.txt should be provided separately for Jarvis purposes?

File: examples/asr/speech_to_text_infer.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            'batch_size': args.batch_size,
+            'normalize_transcripts': args.normalize_text,
+        }
+    )
+    if can_gpu:
+        asr_model = asr_model.cuda()
+    asr_model.eval()
+    labels_map = dict([(i, asr_model.decoder.vocabulary[i]) for i in range(len(asr_model.decoder.vocabulary))])
+    wer = WER(vocabulary=asr_model.decoder.vocabulary)
+    hypotheses = []
```
Comment (titu1994) [inline examples/asr/speech_to_text_infer.py:67]: Is there a particular reason we are doing this manually and not depending on trainer.test()?

File: nemo/collections/nlp/models/token_classification/ner_model.py [hunk 5]
```diff
@@ -54,36 +54,40 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
-        self.model_cfg = cfg
+        # self.model_cfg = cfg
 
         self.tokenizer = get_tokenizer(
             tokenizer_name=cfg.language_model.tokenizer,
             pretrained_model_name=cfg.language_model.pretrained_model_name,
-            vocab_file=cfg.language_model.vocab_file,
-            tokenizer_model=cfg.language_model.tokenizer_model,
+            vocab_file=self.register_artifact(
+                conf_path='language_model.vocab_file', src=cfg.language_model.vocab_file
```
Comment (titu1994) [inline nemo/collections/nlp/models/token_classification/ner_model.py:63]: Lets use the full `config_path=` rather than `conf_path`.

File: nemo/core/classes/common.py [hunk 6]
```diff
@@ -236,34 +241,54 @@ def to_config_file(self, path2yaml_file: str):
...
-    def from_pretrained(cls, name: str):
+    def from_pretrained(cls, model_name: str, refresh_cache: bool = False):
         """
         Instantiates an instance of NeMo from NVIDIA NGC cloud
         Args:
-            name: string key which will be used to find the module. Could be path to local .nemo file.
-
+            model_name: string key which will be used to find the module. Could be path to local .nemo file.
+            refresh_cache: If set to True, then when fetching from clould, this will re-fetch the file
+                from clould even if it is  already found in a cache locally.
```
Comment (titu1994) [inline nemo/core/classes/common.py:270]: fetching from "cloud", below as well. Extra space in between "it is  already found"

File: nemo/core/classes/common.py [hunk 7]
```diff
@@ -236,34 +241,54 @@ def to_config_file(self, path2yaml_file: str):
...
         Returns:
-            A model instance of a class derived from INMModelAPI
+            A model instance of a particular model class
         """
-        pass
+        location_in_the_cloud = None
+        for pretrained_model_info in cls.list_available_models():
```
Comment (titu1994) [inline nemo/core/classes/common.py:275]: This should possibly check for None first and then run the loop

File: nemo/core/classes/modelPT.py [hunk 8]
```diff
@@ -77,6 +85,39 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
+    def register_artifact(self, conf_path: str, src: str):
+        """
+        Register model artifacts with this function. These artifacts (files) will be included inside .nemo file
+        when model.save_to("mymodel.nemo") is called.
+
+        WARNING: If you specified /example_folder/example.txt but ./example.txt exists, then ./example.txt will be used.
+
+        Args:
+            conf_path: config path where artifact is used
+            src: path to the artifact
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:97]: Lets use `config_path` instead of `conf_path`, as mentioned above.

File: nemo/core/classes/modelPT.py [hunk 9]
```diff
@@ -92,11 +133,16 @@ def save_to(self, save_path: str):
         with tempfile.TemporaryDirectory() as tmpdir:
             config_yaml = path.join(tmpdir, _MODEL_CONFIG_YAML)
             model_weights = path.join(tmpdir, _MODEL_WEIGHTS)
+            if hasattr(self, 'artifacts') and self.artifacts is not None:
+                for (conf_path, src) in self.artifacts:
+                    try:
+                        if os.path.exists(src):
+                            shutil.copy2(src, tmpdir)
+                    except:
+                        logging.error(f"Could not copy artifact {src} used in {conf_path}")
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:142]: PEP8, lets use `except Exception:` to catch the general exceptions.

File: nemo/collections/nlp/models/token_classification/ner_model.py (hunk 3 above)
Comment (okuchaiev) [reply nemo/collections/nlp/models/token_classification/ner_model.py:64]: I think if you are not providing vocabulary file when you create your model, then .save_to() should not include any vocabulary files inside .nemo file.

It seems like in this case you are pulling tokenizer directly from HF based on name, then, I think, Jarvis should do exactly the same. @ryanleary does it make sense?

This functionality is generic for the cases when vocabulary and/or model files are indeed specified

File: nemo/utils/cloud.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    """
+    Helper function to download pre-trained weights from the cloud
+    Args:
+        url: (str) URL of storage
+        filename: (str) what to download. The request will be issued to url/filename
+        subfolder: (str) subfolder within cache_dir. The file will be stored in cache_dir/subfolder. Subfolder can
+            be empty
+        cache_dir: (str) a cache directory where to download. If not present, this function will attempt to create it.
+            If None (default), then it will be $HOME/.cache/torch/NeMo
+        referesh_cache: (bool) if True and cached file is present, it will delete it and re-fetch
```
Comment (titu1994) [inline nemo/utils/cloud.py:33]: "refresh"_cache

File: tests/core/test_fileio.py [hunk 11]
```diff
@@ -120,8 +121,10 @@ def test_save_restore_from_nemo_file(self, asr_model):
         with tempfile.NamedTemporaryFile() as fp:
             filename = fp.name
 
-            # Save model.
-            asr_model.save_to(save_path=filename)
+            # Save model (with random artifact).
+            with tempfile.NamedTemporaryFile() as artifact:
+                asr_model.register_artifact(conf_path=None, src=artifact.name)
```
Comment (titu1994) [inline tests/core/test_fileio.py:126]: What happens when config path is None? I see self.artifacts just appends (None, path)?

File: nemo/collections/asr/models/ctc_models.py [hunk 12]
```diff
@@ -35,6 +35,17 @@
 class EncDecCTCModel(ASRModel):
     """Encoder decoder CTC-based models."""
 
+    @classmethod
+    def list_available_models(cls) -> Optional[PretrainedModelInfo]:
```
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:39]: can we support `location` to be passed by the user as a local path? Or do we not want .nemo files to be downloaded manually at all?

File: tests/core/test_fileio.py (hunk 11 above)
Comment (okuchaiev) [reply tests/core/test_fileio.py]: just ignores it, still adds file.

File: nemo/utils/cloud.py (hunk 10 above)
Comment (okuchaiev) [reply nemo/utils/cloud.py]: fixed

File: nemo/core/classes/modelPT.py (hunk 9 above)
Comment (okuchaiev) [reply nemo/core/classes/modelPT.py:142]: changed, but why is "except Exception" better than just "except" ?

File: nemo/core/classes/modelPT.py (hunk 8 above)
Comment (okuchaiev) [reply nemo/core/classes/modelPT.py:97]: changed

File: nemo/core/classes/common.py (hunk 7 above)
Comment (okuchaiev) [reply nemo/core/classes/common.py]: added

File: nemo/core/classes/common.py (hunk 6 above)
Comment (okuchaiev) [reply nemo/core/classes/common.py]: fixed

File: nemo/collections/nlp/models/token_classification/ner_model.py (hunk 5 above)
Comment (okuchaiev) [reply nemo/collections/nlp/models/token_classification/ner_model.py]: changed

File: examples/asr/speech_to_text_infer.py (hunk 4 above)
Comment (okuchaiev) [reply examples/asr/speech_to_text_infer.py:67]: yes: for inference wanted to have PTL-free version. IMHO, nobody will actually run inference with PTL for anything other than publication, so wanted to give "pure" torch variant

File: Jenkinsfile (hunk 2 above)
Comment (okuchaiev) [reply Jenkinsfile]: forgot to remove this. Yes, it is added - L2: ASR Checkponts tests

File: nemo/collections/asr/models/ctc_models.py (hunk 12 above)
Comment (okuchaiev) [reply nemo/collections/asr/models/ctc_models.py:39]: will add it in the next commit

Comment (okuchaiev) [reply nemo/collections/asr/models/ctc_models.py:39]: actually, I'm not exactly sure what you mean. Where will names and description come from?

Yes, the user can download .nemo file from NGC and to "model = ModelClass.restore_from("file_from_NGC.nemo")"

File: examples/asr/speech_to_text_infer.py (hunk 4 above)
Comment (titu1994) [reply examples/asr/speech_to_text_infer.py:67]: Good point

File: nemo/collections/asr/models/ctc_models.py (hunk 12 above)
Comment (titu1994) [reply nemo/collections/asr/models/ctc_models.py:39]: Ah that second part is what I wanted. Very convenient. Nice !

File: nemo/core/classes/modelPT.py (hunk 9 above)
Comment (titu1994) [reply nemo/core/classes/modelPT.py:142]: Its not better but PEP8 has weird rules.
